frontend/src/components/LoginForm.jsx

```diff
@@ -0,0 +1,74 @@
+import React, { useState, useContext } from 'react';
+import { Link, useNavigate } from 'react-router-dom';
+import { UserContext } from '../../UserContext.js';
+
+const LoginForm = () => {
+    const [username, setUsername] = useState('');
+    const [password, setPassword] = useState('');
+    const { updateUser } = useContext(UserContext);
+
+    const navigate = useNavigate();
+
+    const handleLogin = async (e) => {
+        e.preventDefault();
+
+        try {
+            const response = await fetch('http://localhost:3000/users/login', {
+                method: 'POST',
+                headers: {
+                    'Content-Type': 'application/json',
+                },
+                body: JSON.stringify({ username, password }),
+                credentials: 'include',
+            });
+
+            if (response.ok) {
+                const data = await response.json();
+                const loggedInUser = data.user;
+
+                updateUser(loggedInUser);
+
+                navigate('/');
+            } else {
+                const errorData = await response.json();
+                alert(`Login failed: ${errorData.error}`);
+            }
+        } catch (error) {
+            alert('Login failed: ' + error);
```

Comment:
this is probably fine for our project, but in real-life cases it's generally a poor practice to log errors directly to the user. Often times, the error is not meaningful to an end-user and is better suited to be logged to a separate error logging service instead (for consumption by developers)